src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+package pcap
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+)
+
+const concurrentCapturesPerClient = 10
+
+type API struct {
+	bufConf  BufferConf
+	handlers map[string]CaptureHandler
+	agents   AgentMTLS
+	UnimplementedAPIServer
+	id string
+
+	captures               map[string]map[string]*API_CaptureServer
+	captureLock            sync.RWMutex
+	maxConcurrentCaptures  int
+	draining               bool
+	drainTimeout           time.Duration
+	agentCapturesPerVcapID map[string][]*captureReceiver
```

Comment:
This probably needs a discussion: Why do we need to keep track of all streams? The whole point of having channels and contexts is to avoid locks and manual synchronisations. The logic should be designed in a way that allows for synchronisation to happen without explicitly forcing it.

In this case what is probably missing is a way to inform the routines that handle interactions with the upstream agents to terminate once the API is draining. This use-case has already been solved in the pcap-agent (via a channel that is closed once done) so why wasn't the same logic applied here?

Reply:
The idea was to send a clean Stop message. But closing the client side stream would likely have the same effect. Talking to @b1tamara, I now understood the bit with the closed channel and how it would be applied to the pcap-api use case.
Basically you make everything try to read from the done channel (which is never written to), until it closes and then do the cleanup in some local routine that still knows its gRPC stream.
The only bit where we still need to keep track of a number of concurrent captures can be done via atomic counter.

Reply:
We reimplemented the drain logic and we do not need the captures per vcapID anymore.